conformance_tests/tools/debug/src/test_debug.cpp

```diff
@@ -1376,6 +1377,143 @@ void zetDebugReadWriteRegistersTest::run_read_write_registers_test(
+
+    for (auto &stopped_thread : stopped_threads) {
+      uint8_t *mode_values = nullptr;
+      auto register_set_properties = lzt::get_register_set_properties(device);
+      for (auto &register_set : register_set_properties){
+
+        if (register_set.type == ZET_DEBUG_REGSET_TYPE_MODE_FLAGS_INTEL_GPU){
+          auto reg_size_in_bytes = register_set.count * register_set.byteSize;
+          mode_values = new uint8_t[reg_size_in_bytes];
+          ASSERT_EQ(
+              zetDebugReadRegisters(debugSession, stopped_thread,
+                                    ZET_DEBUG_REGSET_TYPE_MODE_FLAGS_INTEL_GPU,
+                                    0, register_set.count, mode_values),
+              ZE_RESULT_SUCCESS);
+
+          uint32_t *uint32_t_values = (uint32_t *)mode_values;
+          LOG_DEBUG << "[Debugger] mode value: %u " << uint32_t_values[0];
+          is_heapless_mode =
+              (uint32_t_values[0] & ZET_DEBUG_MODE_FLAG_HEAPLESS);
+        }
+      }
+
+      if (is_heapless_mode) {
+        for (auto &register_set : register_set_properties){
+          if ((register_set.type ==
+               ZET_DEBUG_REGSET_TYPE_THREAD_SCRATCH_INTEL_GPU) &&
+              (register_set.generalFlags & ZET_DEBUG_REGSET_FLAG_READABLE)) {
+            LOG_DEBUG << "[Debugger] Register set type " << register_set.type
+                      << " is readable";
+            size_t reg_size_in_bytes =
+                register_set.count * register_set.byteSize;
+
+            uint64_t *thread_scratch_reg_values =
+                new uint64_t[reg_size_in_bytes];
+            ASSERT_EQ(zetDebugReadRegisters(
+                          debugSession, stopped_thread,
+                          ZET_DEBUG_REGSET_TYPE_DEBUG_SCRATCH_INTEL_GPU, 0,
+                          register_set.count, thread_scratch_reg_values),
+                      ZE_RESULT_SUCCESS);
+          } else {
+            LOG_INFO << "[Debugger] Register set type " << register_set.type
+                     << " is NOT readable";
+          }
+          if (register_set.generalFlags & ZET_DEBUG_REGSET_FLAG_WRITEABLE) {
+          } else {
+            LOG_INFO << "[Debugger] Register set " << register_set.type
+                     << " type is NOT writeable";
+          }
+        }
+      }
+    }
+
+    lzt::debug_write_memory(debugSession, device_threads, memorySpaceDesc, 1,
+                            kernel_buffer);
+    delete[] kernel_buffer;
+
+    LOG_INFO << "[Debugger] resuming interrupted threads";
+    lzt::debug_resume(debugSession, device_threads);
+    debugHelper.wait();
+
+    std::vector<zet_debug_event_type_t> expectedEvents = {
+        ZET_DEBUG_EVENT_TYPE_MODULE_UNLOAD, ZET_DEBUG_EVENT_TYPE_PROCESS_EXIT};
+
+    if (!check_events(debugSession, expectedEvents)) {
+      FAIL() << "[Debugger] Did not receive expected events";
+    }
+
+    lzt::debug_detach(debugSession);
+    ASSERT_EQ(debugHelper.exit_code(), 0);
+  }
+}
+
+TEST_F(
```

Comment:
Query why cant we use the same test case which @bmyates mentioned in our sync?

Reply:
Then we might need to modify the kernel too. Kernel is not allocating any local memory as such.

Reply:
Any difficulty in modifing the kernel? I hope scratch is forced using IGC_OPTs, so maybe with these options kernel modification might not be required

Reply:
yes please try to use the same test. The kernel should not need to allocate any local memory.  Local memory is for SLM - scratch is different. The compiler will determine if scratch is needed or not based on register usage.  

Try to use same kernel. If scratch isnt allocated then we can try to use IGC opt.  If we still are not getting any spills I can give some suggestions to modify kernel